cpp/src/replace/replace.cu

```diff
@@ -13,26 +16,551 @@
+}
+
+// return the new_value for output column at index `idx`
+template<class T, bool replacement_has_nulls>
+__device__ auto get_new_value(cudf::size_type         idx,
+                           const T* __restrict__ input_data,
+                           const T* __restrict__ values_to_replace_begin,
+                           const T* __restrict__ values_to_replace_end,
+                           const T* __restrict__       d_replacement_values,
+                           cudf::bitmask_type const * __restrict__ replacement_valid)
+   {
+     auto found_ptr = thrust::find(thrust::seq, values_to_replace_begin,
+                                      values_to_replace_end, input_data[idx]);
+     T new_value{0};
+     bool output_is_valid{true};
+
+     if (found_ptr != values_to_replace_end) {
+       auto d = thrust::distance(values_to_replace_begin, found_ptr);
+       new_value = d_replacement_values[d];
+       if (replacement_has_nulls) {
+         output_is_valid = cudf::bit_is_set(replacement_valid, d);
+       }
+     } else {
+       new_value = input_data[idx];
+     }
+     return thrust::make_pair(new_value, output_is_valid);
+   }
+
+  /* --------------------------------------------------------------------------*/
+  /**
+   * @brief Kernel that replaces elements from `output_data` given the following
+   *        rule: replace all `values_to_replace[i]` in [values_to_replace_begin`,
+   *        `values_to_replace_end`) present in `output_data` with `d_replacement_values[i]`.
+   *
+   * @tparam input_has_nulls `true` if output column has valid mask, `false` otherwise
+   * @tparam replacement_has_nulls `true` if replacement_values column has valid mask, `false` otherwise
+   * The input_has_nulls and replacement_has_nulls template parameters allows us to specialize
+   * this kernel for the different scenario for performance without writing different kernel.
+   *
+   * @param[in] input_data Device array with the data to be modified
+   * @param[in] input_valid Valid mask associated with input_data
+   * @param[out] output_data Device array to store the data from input_data
+   * @param[out] output_valid Valid mask associated with output_data
+   * @param[out] output_valid_count #valid in output column
+   * @param[in] nrows # rows in `output_data`
+   * @param[in] values_to_replace_begin Device pointer to the beginning of the sequence
+   * of old values to be replaced
+   * @param[in] values_to_replace_end  Device pointer to the end of the sequence
+   * of old values to be replaced
+   * @param[in] d_replacement_values Device array with the new values
+   * @param[in] replacement_valid Valid mask associated with d_replacement_values
+   *
+   * @returns
+   */
+  /* ----------------------------------------------------------------------------*/
+  template <class T,
+            bool input_has_nulls, bool replacement_has_nulls>
+  __global__
+  void replace_kernel(const T* __restrict__           input_data,
```

Comment:
This should use `column_device_view`s, not raw pointers to the data and bitmasks.

Reply:
Changed.